### PR TITLE
1D Surface FNO Decoder: spectral pressure prediction on arc-length

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Surface pressure p(s) along an airfoil is fundamentally a **1D function of arc-length** s ∈ [0, 1]. The Transolver backbone operates on unstructured 2D mesh nodes with global slice attention — but surface pressure has strong spectral structure: low-frequency components capture the global Cp distribution (lift, drag), while high-frequency components capture the leading-edge suction peak and trailing-edge separation. A **1D Fourier Neural Operator (FNO)** decoder operating in the spectral domain on a uniform arc-length grid should capture this multi-scale structure more naturally than the current SRF (which is a pure MLP on backbone features).

**Key insight:** FNO's spectral convolution layers learn to multiply in frequency space — they can simultaneously model global pressure trends (first ~4 Fourier modes) and local sharp features (higher modes). This is the correct mathematical tool for a smooth-but-sharp 1D function like Cp(s).

**No full FNO architectures have been tried in this programme** — the failed architectures (GNOT, Galerkin, HPT, FactFormer, DeepONet, SIREN) were all different paradigms. 1D FNO as a surface decoder (not a full backbone replacement) is novel here.

**Target metrics:** p_in, p_tan (primary — better spectral decomposition of surface pressure). p_oodc, p_re (secondary — the spectral representation may generalize better across geometries/Re).

**Reference:** Li et al. (2021) "Fourier Neural Operator for Parametric PDEs" — showed FNO captures multi-scale PDE solutions efficiently via learnable spectral filters. Patel et al. (2024) "On the Fourier Neural Operator for Real-Time Surface Pressure Prediction" — directly applied FNO to surface pressure.

## Instructions

Add a `--surface_fno` flag to `cfd_tandemfoil/train.py`. When active, replace the SRF head with a 1D FNO decoder operating on surface arc-length.

### 1. Add flags
```python
parser.add_argument('--surface_fno', action='store_true',
    help='Replace SRF with 1D FNO decoder on arc-length parameterized surface')
parser.add_argument('--surface_fno_modes', type=int, default=16,
    help='Number of Fourier modes to keep (default: 16)')
parser.add_argument('--surface_fno_width', type=int, default=64,
    help='FNO hidden channel width (default: 64)')
parser.add_argument('--surface_fno_layers', type=int, default=4,
    help='Number of FNO layers (default: 4)')
parser.add_argument('--surface_fno_grid', type=int, default=128,
    help='Uniform grid resolution for surface (default: 128)')
```

### 2. Arc-length parameterization

Sort surface nodes by angle from centroid (or cumulative Euclidean distance) to get an ordered 1D curve. Then compute arc-length fraction s ∈ [0, 1]:

```python
def parameterize_surface(surface_xy, surface_mask):
    """
    surface_xy: [N_surf, 2] — coordinates of surface nodes
    Returns: arc_length [N_surf] in [0, 1], sort_indices [N_surf]
    """
    # Sort by angle from centroid
    centroid = surface_xy.mean(dim=0)
    angles = torch.atan2(surface_xy[:, 1] - centroid[1],
                         surface_xy[:, 0] - centroid[0])
    sort_idx = angles.argsort()
    sorted_xy = surface_xy[sort_idx]

    # Cumulative arc-length
    diffs = torch.diff(sorted_xy, dim=0)
    segment_lengths = diffs.norm(dim=1)
    arc = torch.cat([torch.zeros(1, device=segment_lengths.device),
                     segment_lengths.cumsum(0)])
    arc = arc / (arc[-1] + 1e-8)  # normalize to [0, 1]
    return arc, sort_idx
```

For **tandem** samples: parameterize each foil separately (fore: foil_id==1, aft: foil_id==2). Apply FNO to each independently, then concatenate results.

### 3. Interpolate to uniform grid

```python
def interpolate_to_grid(features, arc_length, n_grid=128):
    """
    features: [N_surf, D] — backbone embeddings at surface nodes
    arc_length: [N_surf] in [0, 1]
    Returns: grid_features [n_grid, D]
    """
    # Uniform grid points
    grid_s = torch.linspace(0, 1, n_grid, device=features.device)
    # Linear interpolation using torch.searchsorted
    idx = torch.searchsorted(arc_length, grid_s).clamp(1, len(arc_length) - 1)
    s0, s1 = arc_length[idx - 1], arc_length[idx]
    w = ((grid_s - s0) / (s1 - s0 + 1e-8)).unsqueeze(1)
    grid_feat = (1 - w) * features[idx - 1] + w * features[idx]
    return grid_feat
```

### 4. 1D FNO layer (pure PyTorch)

```python
class SpectralConv1d(nn.Module):
    """1D Fourier layer: multiply in frequency domain"""
    def __init__(self, in_channels, out_channels, modes):
        super().__init__()
        self.modes = modes
        # Complex weights for spectral multiplication
        self.weights = nn.Parameter(
            torch.randn(in_channels, out_channels, modes, 2) * 0.02
        )

    def forward(self, x):
        # x: [B, C, N] (batch, channels, spatial)
        B, C, N = x.shape
        x_ft = torch.fft.rfft(x, dim=-1)  # [B, C, N//2+1]
        # Multiply first 'modes' frequencies by learned weights
        weights_c = torch.view_as_complex(self.weights)  # [C_in, C_out, modes]
        out_ft = torch.zeros(B, self.weights.shape[1], N // 2 + 1,
                             dtype=torch.cfloat, device=x.device)
        out_ft[:, :, :self.modes] = torch.einsum(
            'bci,ioc->boc', x_ft[:, :, :self.modes], weights_c
        )
        return torch.fft.irfft(out_ft, n=N, dim=-1)  # [B, C_out, N]


class FNOLayer1d(nn.Module):
    def __init__(self, width, modes):
        super().__init__()
        self.spectral = SpectralConv1d(width, width, modes)
        self.linear = nn.Conv1d(width, width, 1)  # local skip
        self.norm = nn.InstanceNorm1d(width)

    def forward(self, x):
        # x: [B, C, N]
        return self.norm(x + F.gelu(self.spectral(x) + self.linear(x)))


class SurfaceFNO(nn.Module):
    def __init__(self, in_dim, width=64, modes=16, n_layers=4, out_dim=3):
        super().__init__()
        self.lift = nn.Linear(in_dim, width)
        self.layers = nn.ModuleList([FNOLayer1d(width, modes) for _ in range(n_layers)])
        self.proj = nn.Linear(width, out_dim)
        # Zero-init final projection for safe startup
        nn.init.zeros_(self.proj.weight)
        nn.init.zeros_(self.proj.bias)

    def forward(self, x):
        # x: [B, N_grid, in_dim]
        x = self.lift(x)            # [B, N_grid, width]
        x = x.permute(0, 2, 1)     # [B, width, N_grid] for Conv1d
        for layer in self.layers:
            x = layer(x)
        x = x.permute(0, 2, 1)     # [B, N_grid, width]
        return self.proj(x)         # [B, N_grid, out_dim]
```

### 5. Integration into training

When `--surface_fno` is active:
1. After backbone → extract surface node embeddings `h_surf`
2. For each foil in sample: parameterize surface → interpolate to uniform grid → run FNO
3. Interpolate FNO output back to original mesh node positions (inverse of step 2)
4. Use FNO output as the surface prediction (replace SRF head)
5. Note: **omit `--surface_refine`** since FNO replaces it

### 6. Per-foil processing for tandem

```python
if is_tandem:
    # Fore foil
    fore_arc, fore_sort = parameterize_surface(xy[fore_surf_mask])
    fore_grid = interpolate_to_grid(h[fore_surf_mask][fore_sort], fore_arc, n_grid)
    fore_out = self.surface_fno(fore_grid.unsqueeze(0))  # [1, N_grid, 3]
    # Interpolate back to original fore surface nodes
    fore_preds = interpolate_from_grid(fore_out.squeeze(0), fore_arc, fore_sort)

    # Aft foil (same process)
    aft_arc, aft_sort = parameterize_surface(xy[aft_surf_mask])
    aft_grid = interpolate_to_grid(h[aft_surf_mask][aft_sort], aft_arc, n_grid)
    aft_out = self.surface_fno(aft_grid.unsqueeze(0))
    aft_preds = interpolate_from_grid(aft_out.squeeze(0), aft_arc, aft_sort)
else:
    # Single foil
    arc, sort_idx = parameterize_surface(xy[surf_mask])
    grid = interpolate_to_grid(h[surf_mask][sort_idx], arc, n_grid)
    fno_out = self.surface_fno(grid.unsqueeze(0))
    preds = interpolate_from_grid(fno_out.squeeze(0), arc, sort_idx)
```

### Run commands

```bash
# Seed 42
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent frieren --wandb_name "frieren/surface-fno-s42" --wandb_group surface-fno \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep \
  --residual_prediction --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only \
  --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity \
  --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --surface_fno --surface_fno_modes 16 --surface_fno_width 64 \
  --surface_fno_layers 4 --surface_fno_grid 128

# Seed 73 (same flags, --seed 73, --wandb_name "frieren/surface-fno-s73")
```

Note: **omit `--surface_refine`** since the FNO replaces it.

### Logging
```python
wandb.log({
    'train/fno_out_norm': fno_out.norm().item(),
    'train/fno_modes_energy': (fno_weights_complex.abs() ** 2).sum().item(),
})
```

## Baseline

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| **p_in** | **11.872** | < 11.872 |
| p_oodc | 7.459 | < 7.459 |
| **p_tan** | **26.319** | < 26.319 |
| **p_re** | **6.229** | < 6.229 |

W&B baseline: aycq1m8m (seed 42, epoch 155), 9sk276v6 (seed 73, epoch 156)

Reproduce:
```bash
cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output \
  --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Summary

**Negative result** — the 1D Surface FNO decoder significantly underperforms the baseline across all metrics. Surface pressure MAE is worse by 15–29% on every split.

### Metrics

| Metric | Seed 42 | Seed 73 | **2-seed avg** | Baseline | Δ |
|--------|---------|---------|---------------|----------|---|
| **p_in** | 14.5 | 14.2 | **14.35** | 11.872 | +2.48 (+20.9%) ❌ |
| **p_tan** | 34.4 | 33.5 | **33.95** | 26.319 | +7.63 (+29.0%) ❌ |
| **p_oodc** | 8.9 | 8.6 | **8.75** | 7.459 | +1.29 (+17.3%) ❌ |
| **p_re** | 7.0 | 7.3 | **7.15** | 6.229 | +0.92 (+14.8%) ❌ |

**In-distribution (seed avg, val_in_dist):** surf_Ux=3.41, surf_Uy=1.14, surf_p=14.35, vol_Ux=0.59, vol_Uy=0.22, vol_p=13.63

**Best epochs:** S42=122 (val/loss=0.6855), S73=121 (val/loss=0.6808) — both hitting 180-min wall clock limit

### Train commands

```bash
# Seed 42
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent frieren --wandb_name "frieren/surface-fno-s42" --wandb_group surface-fno \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep \
  --residual_prediction --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only \
  --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity \
  --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --surface_fno --surface_fno_modes 16 --surface_fno_width 64 \
  --surface_fno_layers 4 --surface_fno_grid 128

# Seed 73: same with --seed 73 --wandb_name "frieren/surface-fno-s73"
```

### Peak memory

S42: 40.2 GB, S73: 40.9 GB (slight increase vs baseline ~40.1 GB due to FNO parameters + grid buffers)

### W&B run IDs

- Seed 42: `40gf9jfw`
- Seed 73: `fdzyjdc9`
- W&B group: `surface-fno`

### What happened

The 1D FNO decoder failed to improve over the SRF baseline. Several likely causes:

1. **FNO replaces SRF entirely**: The baseline SRF is an established, working refinement head. Replacing it with FNO removes a proven component. By contrast, the FNO's spectral filters — while theoretically better suited for smooth-but-sharp 1D functions — require arc-length interpolation that may lose critical spatial detail.

2. **Arc-length interpolation loses fine structure**: Converting unstructured surface nodes to a uniform 128-point arc-length grid and back introduces interpolation error at high-curvature regions (leading-edge suction peak, trailing edge). These are exactly the regions where pressure is hardest to predict and most important for accuracy.

3. **Tandem p_tan badly hurt (+29%)**: For tandem samples, the FNO processes fore and aft foils independently. The aft foil's hidden state captures wake interference implicitly through backbone attention, but the 1D FNO — operating only on arc-length-sorted surface features of that foil — may not preserve the wake coupling information that the SRF sees more directly.

4. **Limited convergence (122 epochs vs 155 baseline)**: Both runs hit the 180-min wall clock at epochs 121–122. The baseline trained to epoch 155 with cosine T_max=150. At epoch 122 the val losses were still declining (val_in_dist ~0.44–0.45), suggesting more epochs could reduce the gap — but the trajectories don't suggest catching up to baseline.

5. **Zero-init delay**: Zero-initializing the FNO projection layer ensures stable startup but means FNO only starts meaningfully contributing after many epochs. The SRF presumably learns useful corrections earlier.

### Suggested follow-ups

1. **FNO + SRF in combination** (`--surface_refine --surface_fno`): Use FNO as a spectral residual *on top of* SRF rather than as a replacement. This would test whether FNO adds value incrementally rather than being a full alternative.

2. **Higher grid resolution** (`--surface_fno_grid 256`): Finer grid reduces interpolation error at sharp leading-edge features. Could recover some of the accuracy loss from coarse interpolation.

3. **FNO as volume head**: Rather than surface pressure, try applying FNO-style spectral processing to the volume field (e.g., in the decoder, for wake regions), where the 2D structure might benefit from spectral decomposition.

4. **Fewer Fourier modes** (`--surface_fno_modes 8`): The 16 modes may be overfitting with limited data; fewer modes would enforce smoother spectral representations.

5. **Learnable arc-length**: Instead of angle-from-centroid ordering, learn a surface traversal order via differentiable sorting or use a pre-computed signed arc-length from the mesh topology.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2371 because ADVISOR: Closing — all metrics 15-29% above baseline. The 1D FNO decoder is a clear dead end.

Your analysis is thorough and correct. The core failure modes:
1. **Arc-length interpolation destroys critical fine structure** at LE/TE — exactly where pressure prediction matters most. This is fundamental to the FNO-on-surface approach.
2. **Per-foil 1D processing severs tandem coupling** — the backbone encodes inter-foil wake physics into the hidden state, but the FNO can't preserve it when processing each foil's arc-length sequence independently.
3. **Compute overhead reduced effective training** to 122 epochs vs 155 for baseline.

This confirms: spectral decomposition on 1D surface arc-length is not the right inductive bias. The surface pressure distribution has sharp features (LE suction peak, TE Kutta condition) that require spatial rather than spectral resolution.

New assignment incoming..
